arch/xtensa/src/esp32s3/esp32s3_spi_timing.c

```diff
@@ -238,129 +349,1232 @@ static void IRAM_ATTR set_flash_clock(uint8_t spi_num, uint32_t freqdiv)
+ *
+ * Input Parameters:
+ *   config             - tuning timing configuration pointer
+ *   consecutive_length - consecutive length
+ *   end                - last success point
+ *
+ * Returned Value:
+ *   Returns the STR best point.
+ *
+ ****************************************************************************/
+
+#if MSPI_TIMING_FLASH_STR_MODE || MSPI_TIMING_PSRAM_STR_MODE
+static uint32_t select_best_tuning_config_str(tuning_config_s *config,
+                          uint32_t consecutive_length, uint32_t end)
+{
+#if (ESP32S3_SPI_TIMING_CORE_CLK == 120 || ESP32S3_SPI_TIMING_CORE_CLK == 240)
+  minfo("DO NOT USE FOR MASS PRODUCTION! Timing parameters may be updated");
+
+  /* STR best point scheme */
+
+  uint32_t best_point;
+  if (consecutive_length <= 2 || consecutive_length >= 5)
+    {
+      /* tuning is FAIL, select default point, and generate a warning */
+
+      best_point = config->default_config_id;
+      minfo("tuning fail, best point is fallen back to index %d\n",
+            best_point);
+    }
+  else
+    {
+      /* consecutive length :  3 or 4 */
+
+      best_point = end - consecutive_length / 2;
+      minfo("tuning success, best point is index %d\n", best_point);
+    }
+
+  return best_point;
+#else
+  /* won't reach here */
+
+  abort();
+#endif
+}
+#endif
+
+/****************************************************************************
+ * Name: select_best_tuning_config_dtr
+ *
+ * Description:
+ *   Select the DTR best point
+ *
+ * Input Parameters:
+ *   config             - tuning timing configuration pointer
+ *   consecutive_length - consecutive length
+ *   end                - last success point
+ *
+ * Returned Value:
+ *   Returns the DTR best point.
+ *
+ ****************************************************************************/
+
+#if MSPI_TIMING_FLASH_DTR_MODE || MSPI_TIMING_PSRAM_DTR_MODE
+static uint32_t select_best_tuning_config_dtr(tuning_config_s *config,
+                           uint32_t consecutive_length, uint32_t end)
+{
+#if (ESP32S3_SPI_TIMING_CORE_CLK == 160)
+
+  /* Core clock 160M DTR best point scheme */
+
+  uint32_t best_point;
+
+  /* These numbers will probably be same on other chips,
+   * if this version of algorithm is utilised
+   */
+
+  if (consecutive_length <= 2 || consecutive_length >= 6)
+    {
+      /* tuning is FAIL, select default point, and generate a warning */
+
+      best_point = config->default_config_id;
+      minfo("tuning fail, best point is fallen back to index %d\n",
+            best_point);
+    }
+  else if (consecutive_length <= 4)
+    {
+      /* consecutive length :  3 or 4 */
+
+      best_point = end - 1;
+      minfo("tuning success, best point is index %d\n", best_point);
+    }
+  else
+    {
+      /* consecutive point list length equals 5 */
+
+      best_point = end - 2;
+      minfo("tuning success, best point is index %d\n", best_point);
+    }
+
+  return best_point;
+#else
+  /* won't reach here */
+
+  abort();
```

Comment:
I'm not sure that this will work from kernel

Reply:
I've confirmed `abort()` function will block system here.